src/Reflection/FunctionVariant.php

```diff
@@ -51,4 +52,15 @@ public function getReturnType(): Type
 		return $this->returnType;
 	}
 
+	/**
+	 * @return static
+	 */
+	public function flattenConditionalsInReturnType(): self
+	{
+		$result = clone $this;
```

Comment:
I dislike `clone`, please create new instance explicitly with `new self`.

Reply:
Hmm, that's disappointing. Given that these classes are marked `@api` and the constructor isn't final, constructing an instance of `static` seems impossible.

Or would you accept a version that returns `$this` if `self::class !== static::class`?

Reply:
Give me a moment, I'll think of something 😊

Reply:
Yeah, these classes should ideally be `final`, there's no reason to extend them. I've marked a new todo for 2.0 when the time comes - class has either to be `@api` or final (it can be both).

Reply:
Maybe we can mark them `@final` in the meantime?